### ISSUE TITLE
[BUG] Dask groupby-aggregations with a list of agg funcs with reset_index causes multi-index related error

### ISSUE THREAD

**Issue by beckernick** [rapidsai/cudf#3904]
Groupby-aggregations with a list of agg funcs passed cause multi-index issues with reset_index in dask-cudf.

@thomcom is looking into this. Example below (courtesy of @VibhuJawa ):

```python
import dask_cudf
import cudf
​
df = dask_cudf.from_cudf(cudf.DataFrame({'a':[1,2,3,4],'b':[5,1,2,5]}),npartitions=1)
df_groups = df.groupby(['a']).agg({'b': ['count']}).reset_index().compute()
/opt/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/dataframe.py:380: UserWarning: Columns may not be added to a DataFrame using a new attribute name. A new attribute will be created: 'multi_cols'
  UserWarning,
/opt/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/dataframe.py:380: UserWarning: Columns may not be added to a DataFrame using a new attribute name. A new attribute will be created: 'multi_cols'
  UserWarning,
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-1c92d29d718e> in <module>
      3 
      4 df = dask_cudf.from_cudf(cudf.DataFrame({'a':[1,2,3,4],'b':[5,1,2,5]}),npartitions=1)
----> 5 df_groups = df.groupby(['a']).agg({'b': ['count']}).reset_index().compute()

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/base.py in compute(self, **kwargs)
    163         dask.base.compute
    164         """
--> 165         (result,) = compute(self, traverse=False, **kwargs)
    166         return result
    167 

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/base.py in compute(*args, **kwargs)
    434     keys = [x.__dask_keys__() for x in collections]
    435     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 436     results = schedule(dsk, keys, **kwargs)
    437     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    438 

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/local.py in get_sync(dsk, keys, **kwargs)
    525     """
    526     kwargs.pop("num_workers", None)  # if num_workers present, remove it
--> 527     return get_async(apply_sync, 1, dsk, keys, **kwargs)
    528 
    529 

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/local.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    469             # Seed initial tasks into the thread pool
    470             while state["ready"] and len(state["running"]) < num_workers:
--> 471                 fire_task()
    472 
    473             # Main loop, wait on tasks to finish, insert new ones

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/local.py in fire_task()
    464                         pack_exception,
    465                     ),
--> 466                     callback=queue.put,
    467                 )
    468 

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/local.py in apply_sync(func, args, kwds, callback)
    514 def apply_sync(func, args=(), kwds={}, callback=None):
    515     """ A naive synchronous version of apply_async """
--> 516     res = func(*args, **kwds)
    517     if callback is not None:
    518         callback(res)

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    225         failed = False
    226     except BaseException as e:
--> 227         result = pack_exception(e, dumps)
    228         failed = True
    229     return key, result, failed

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    220     try:
    221         task, data = loads(task_info)
--> 222         result = _execute_task(task, data)
    223         id = get_id()
    224         result = dumps((result, id))

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/core.py in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/optimization.py in __call__(self, *args)
    980         if not len(args) == len(self.inkeys):
    981             raise ValueError("Expected %d args, got %d" % (len(self.inkeys), len(args)))
--> 982         return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
    983 
    984     def __reduce__(self):

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/core.py in get(dsk, out, cache)
    147     for key in toposort(dsk):
    148         task = dsk[key]
--> 149         result = _execute_task(task, cache)
    150         cache[key] = result
    151     result = _execute_task(out, cache)

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/core.py in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/utils.py in apply(func, args, kwargs)
     27 def apply(func, args, kwargs=None):
     28     if kwargs:
---> 29         return func(*args, **kwargs)
     30     else:
     31         return func(*args)

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/core.py in apply_and_enforce(*args, **kwargs)
   4979             return meta
   4980         if is_dataframe_like(df):
-> 4981             check_matching_columns(meta, df)
   4982             c = meta.columns
   4983         else:

/opt/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/utils.py in check_matching_columns(meta, actual)
    669 def check_matching_columns(meta, actual):
    670     # Need nan_to_num otherwise nan comparison gives False
--> 671     if not np.array_equal(np.nan_to_num(meta.columns), np.nan_to_num(actual.columns)):
    672         extra = actual.columns.difference(meta.columns).tolist()
    673         missing = meta.columns.difference(actual.columns).tolist()

<__array_function__ internals> in nan_to_num(*args, **kwargs)

TypeError: no implementation found for 'numpy.nan_to_num' on types that implement __array_function__: [<class 'cudf.core.multiindex.MultiIndex'>]
```

**Comment by rjzamora** [comment]
@thomcom -  I just walked through the trace here a few (hundred) times.  I don't have a solution in place, but I am fairly confident that the fix is to add something like an `__array_function__` definition to `cudf.MultiIndex`.  Dask assumes it can perform a numpy-based operation on a MultiIndex within `check_matching_columns`. When I comment out this check, the `compute()` returns as expected.

The fact that the `meta` and `actual` cudf-dataframes seem to return different types for the `column` property is strange, but not the cause of the error here.  The reason for this difference is hard to follow, but is essentially due the fact that dask is constructing a dataframe by passing the `cudf.DataFrame` constructor an `OrderedDict` with a tuple `('b', 'count')` as the only key.  The `cudf.DataFrame.__init__` method ulimately uses `_init_from_dict_like` to initialize the DataFrame, which in-turn uses `insert` to add a `('b', 'count')` key to `self._data` (i.e. `self._data[name] = value`). 

When you execute `actual.columns`,  the `cudf.DataFrame.columns` property is  simply returning `pd.Index(self._data.keys(), name=name)`.  In this case, `self._data.keys()` includes a tuple, so pandas converts the `pd.Index` into a `pd.MultiIndex`.  The reason that `meta.columns` returns a `cudf.MultiIndex` object instead is because that object is an *empty* cudf dataframe (by disign). 
 In order to create an *empty* dataframe, the `make_meta_cudf` dispatch function uses `.head(0)` (equivalent to `.iloc[:0]`).  For whatever reason,  using `iloc` to take a view results in `columns` being converted to the "correct" type.  You can confirm this by checking the type of `actual.iloc[:].columns` (rather than `actual.columns`).

I haven't had time to work out a fix, but hopefully this information helps a bit.

**Comment by thomcom** [comment]
Thanks for this thorough anaysis @rjzamora!

**Comment by thomcom** [comment]
Can you point out the loc where dask constructs the `cudf.DataFrame` with an `OrderedDict` of `('b', 'count')`? @rjzamora

**Comment by rjzamora** [comment]
> Can you point out the loc where dask constructs the cudf.DataFrame with an OrderedDict of ('b', 'count')? @rjzamora

Yup - I believe it is happening in [in `_agg_finalize`](https://github.com/dask/dask/blob/82c0c752c80b1c3b994ba04693633664a0402fc1/dask/dataframe/groupby.py#L909).  Note the `return type(df)(result)` return statement.

**Comment by randerzander** [comment]
Resolved by #3931

**Comment by beckernick** [comment]
This was resolved but is now presenting again in the distributed scheduler scenario. Cross-linking on #4030
